integration/cluster/cluster.go

```diff
@@ -0,0 +1,348 @@
+package cluster
+
+import (
+	"errors"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"net/http/httptest"
+	"net/url"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/grafana/dskit/multierror"
+	"github.com/prometheus/client_golang/prometheus"
+
+	"github.com/grafana/loki/pkg/loki"
+	"github.com/grafana/loki/pkg/util/cfg"
+)
+
+var wrapRegistryOnce sync.Once
+
+func wrapRegistry() {
+	wrapRegistryOnce.Do(func() {
+		prometheus.DefaultRegisterer = &wrappedRegisterer{Registerer: prometheus.DefaultRegisterer}
+	})
+}
+
+type wrappedRegisterer struct {
+	prometheus.Registerer
+}
+
+func (w *wrappedRegisterer) Register(collector prometheus.Collector) error {
+	if err := w.Registerer.Register(collector); err != nil {
+		var aErr prometheus.AlreadyRegisteredError
+		if errors.As(err, &aErr) {
+			return nil
+		}
+		return err
+	}
+	return nil
+}
+
+func (w *wrappedRegisterer) MustRegister(collectors ...prometheus.Collector) {
+	for _, c := range collectors {
+		if err := w.Register(c); err != nil {
+			panic(err.Error())
+		}
+	}
+}
+
+type Cluster struct {
+	sharedPath string
+	components []*Component
+	waitGroup  sync.WaitGroup
+}
+
+func New() *Cluster {
+	wrapRegistry()
+	sharedPath, err := ioutil.TempDir("", "loki-shared-data")
```

Comment:
The `ioutil` package is somewhat deprecated: https://pkg.go.dev/io/ioutil#pkg-overview.

```suggestion
	sharedPath, err := os.MkdirTemp("", "loki-shared-data")
```